File: src/main/java/io/veridia/jsonlogic/helpers/ToDouble.java
package io.veridia.jsonlogic.helpers;

public class ToDouble {
    public static double eval(Object o) {
        if (o == null) return 0;

        if (o instanceof Boolean) return ((Boolean) o) ? 1 : 0;

        if (o instanceof Number) return ((Number) o).doubleValue();

        if (o instanceof String) {
            String s = (String) o;
            try {
                if (s.trim().isEmpty()) {
                    s = "0";
                }

                return Double.parseDouble(s);
            }
            catch (NumberFormatException e) {
                return 0;
            }
        }

        return 0;
    }
}

File: src/main/scala/io/veridia/jsonlogic/helpers/ToDouble.scala
package io.veridia.jsonlogic.helpers

object ToDouble:
  def eval(o: Any): Double = o match
    case null => 0d
    case b: java.lang.Boolean => if b.booleanValue() then 1d else 0d
    case n: Number => n.doubleValue()
    case s: String =>
      val text = if s.isBlank then "0" else s
      try text.toDouble
      catch case _: NumberFormatException => 0d
    case _ => 0d
